app/cluster/state/etcd.go

```diff
@@ -0,0 +1,289 @@
+	workspacesRequestsKeyPattern = `/%s/server/%s/workspaces` // /<releaseName>/server/<serverIndex>/workspaces
+)
+
+var _ cluster.ModeProvider = &ETCDManager{}
+
+type ETCDConfig struct {
+	Namespace            string
+	ServerIndex          string
+	Endpoints            []string
+	dialKeepAliveTime    time.Duration
+	dialKeepAliveTimeout time.Duration
+	etcdWatchTimeout     time.Duration
+	etcdGetTimeout       time.Duration
+	dialTimeout          time.Duration
+}
+
+type modeRequestValue struct {
+	Mode   servermode.Mode `json:"mode"`
+	AckKey string          `json:"ack_key"`
+}
+
+type modeAckValue struct {
+	Status servermode.Mode `json:"status"`
+}
+
+type workspacesRequestsValue struct {
+	Workspaces string `json:"workspaces"` // comma separated workspaces
+	AckKey     string `json:"ack_key"`
+}
+
+type workspacesAckValue struct {
+	Status string `json:"status"`
+}
+
+func EnvETCDConfig() *ETCDConfig {
+	endpoints := strings.Split(config.GetEnv("ETCD_HOSTS", "127.0.0.1:2379"), `,`)
+	namespace := config.GetKubeNamespace()
+	serverIndex := config.GetInstanceID()
+
+	envConfigOnce.Do(func() {
+		config.RegisterDurationConfigVariable(time.Duration(15), &etcdGetTimeout, false, time.Second, "etcd.getTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(3), &etcdWatchTimeout, false, time.Second, "etcd.watchTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(30), &keepaliveTime, false, time.Second, "etcd.keepaliveTime")
+		config.RegisterDurationConfigVariable(time.Duration(10), &keepaliveTimeout, false, time.Second, "etcd.keepaliveTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(20), &dialTimeout, false, time.Second, "etcd.dialTimeout")
+	})
+
+	return &ETCDConfig{
+		Endpoints:            endpoints,
+		Namespace:            namespace,
+		ServerIndex:          serverIndex,
+		etcdWatchTimeout:     etcdWatchTimeout,
+		dialTimeout:          dialTimeout,
+		dialKeepAliveTime:    keepaliveTime,
+		dialKeepAliveTimeout: keepaliveTimeout,
+		etcdGetTimeout:       etcdGetTimeout,
+	}
+}
+
+type ETCDManager struct {
+	Config  *ETCDConfig
+	Client  *clientv3.Client
+	once    sync.Once
+	initErr error
+}
+
+func (manager *ETCDManager) init() error {
+	manager.once.Do(func() {
+		cli, err := clientv3.New(clientv3.Config{
+			Endpoints:            manager.Config.Endpoints,
+			DialTimeout:          manager.Config.dialTimeout,
+			DialKeepAliveTime:    manager.Config.dialKeepAliveTime,
+			DialKeepAliveTimeout: manager.Config.dialKeepAliveTimeout,
+			DialOptions: []grpc.DialOption{
+				grpc.WithBlock(), // block until the underlying connection is up
+			},
+		})
+		if err != nil {
+			manager.initErr = err
+			return
+		}
+		manager.Client = cli
+	})
+
+	return manager.initErr
+}
+
+// Ping ensures the connection to etcd is alive
+func (manager *ETCDManager) Ping() error {
+	if err := manager.init(); err != nil {
+		return err
+	}
+
+	_, err := manager.Client.Cluster.MemberList(context.Background())
```

Comment:
Why did we choose to use the MemberList function to determine the health of the cluster? I have seen few repos using a get call on any random key for the same functionality

Reply:
No reason in particular. I look inside the etcd implementation and looked that this is called quite frequently for coordination. 

No strong point on this, I can change it if you want.

Reply:
Just wanted to know the idea behind it. Anything that helps establish that there is a connection should be good enough